### PR TITLE
Convert loops dependent on bandlimit in JAX quadrature rules to functional equivalents

### PR DESCRIPTION
**Pull request by matt-graham**
Relates to #317 

Currently the JAX implementations of the quadrature rules (computations of weights for each $\theta$) used in the forward transform, use native Python for loops for some of the sampling schemes. These loops have a number of iterations which scales with the bandlimit `L`. When JAX traces out these functions as part of its just-in-time compilation transform, the loops are unrolled, resulting in traced JAX expressions (jaxprs) which increase in size as `L` is increased, and an accompanying increase in compilation time due to the larger expression tree / graph being compiled.

[JAX offers functional 'structured' control-flow primitives to side step this issue](https://docs.jax.dev/en/latest/control-flow.html#structured-control-flow-primitives). This PR replaces the native Python for loops in the JAX quadrature rules with an equivalent implementation in terms of [the `jax.lax.fori_loop` function](https://docs.jax.dev/en/latest/_autosummary/jax.lax.fori_loop.html#jax.lax.fori_loop). The changes should not change numerics (beyond floating point differences due to slight rearrangements of ordering of operations) and also not have a significant effect on run-time performance, with the benefit mainly in reduced compilation times. This is borne out in benchmarking (copying results from https://github.com/astro-informatics/s2fft/issues/317#issuecomment-3916111784):

On current `main` (cd8db48)

```
forward
(method: jax, L: 16, L_lower: 0, sampling: mw, spin: 0, L_to_nside_ratio: 2, reality: True, spmd: False, n_iter: None): 
    min(run times): 0.00037s, max(run times): 0.00050s, compile time:    0.81s, peak memory: 8.7e+03B, max(abs(error)): 2.9e-14, floating point ops: 4.3e+05, mem access: 1.2e+06B
(method: jax, L: 32, L_lower: 0, sampling: mw, spin: 0, L_to_nside_ratio: 2, reality: True, spmd: False, n_iter: None): 
    min(run times):  0.0015s, max(run times):  0.0016s, compile time:    0.97s, peak memory: 8.4e+03B, max(abs(error)): 8.1e-14, floating point ops: 2.0e+06, mem access: 5.1e+06B
(method: jax, L: 64, L_lower: 0, sampling: mw, spin: 0, L_to_nside_ratio: 2, reality: True, spmd: False, n_iter: None): 
    min(run times):  0.0094s, max(run times):   0.011s, compile time:     1.3s, peak memory: 8.4e+03B, max(abs(error)): 2.6e-13, floating point ops: 9.3e+06, mem access: 2.1e+07B
(method: jax, L: 128, L_lower: 0, sampling: mw, spin: 0, L_to_nside_ratio: 2, reality: True, spmd: False, n_iter: None): 
    min(run times):   0.076s, max(run times):   0.077s, compile time:     2.0s, peak memory: 8.4e+03B, max(abs(error)): 1.9e-12, floating point ops: 4.2e+07, mem access: 9.3e+07B
(method: jax, L: 256, L_lower: 0, sampling: mw, spin: 0, L_to_nside_ratio: 2, reality: True, spmd: False, n_iter: None): 
    min(run times):    0.94s, max(run times):    0.94s, compile time:     4.4s, peak memory: 8.4e+03B, max(abs(error)): 8.1e-12, floating point ops: 1.9e+08, mem access: 3.7e+08B
(method: jax, L: 512, L_lower: 0, sampling: mw, spin: 0, L_to_nside_ratio: 2, reality: True, spmd: False, n_iter: None): 
    min(run times):     9.7s, max(run times):     9.9s, compile time:     16.s, peak memory: 8.4e+03B, max(abs(error)): 3.8e-11, floating point ops: 8.3e+08, mem access: 1.5e+09B
```

On `mmg/jaxify-quadrature-loops` branch (5ef1b85) - that is changes in this PR

```
forward
(method: jax, L: 16, L_lower: 0, sampling: mw, spin: 0, L_to_nside_ratio: 2, reality: True, spmd: False, n_iter: None): 
    min(run times): 0.00038s, max(run times): 0.00049s, compile time:    0.70s, peak memory: 8.7e+03B, max(abs(error)): 1.5e-14, floating point ops: 4.3e+05, mem access: 1.2e+06B
(method: jax, L: 32, L_lower: 0, sampling: mw, spin: 0, L_to_nside_ratio: 2, reality: True, spmd: False, n_iter: None): 
    min(run times):  0.0015s, max(run times):  0.0015s, compile time:    0.77s, peak memory: 8.3e+03B, max(abs(error)): 1.3e-13, floating point ops: 2.0e+06, mem access: 5.1e+06B
(method: jax, L: 64, L_lower: 0, sampling: mw, spin: 0, L_to_nside_ratio: 2, reality: True, spmd: False, n_iter: None): 
    min(run times):   0.011s, max(run times):   0.011s, compile time:    0.71s, peak memory: 8.4e+03B, max(abs(error)): 3.6e-13, floating point ops: 9.3e+06, mem access: 2.1e+07B
(method: jax, L: 128, L_lower: 0, sampling: mw, spin: 0, L_to_nside_ratio: 2, reality: True, spmd: False, n_iter: None): 
    min(run times):   0.068s, max(run times):   0.070s, compile time:    0.78s, peak memory: 8.3e+03B, max(abs(error)): 1.6e-12, floating point ops: 4.2e+07, mem access: 9.3e+07B
(method: jax, L: 256, L_lower: 0, sampling: mw, spin: 0, L_to_nside_ratio: 2, reality: True, spmd: False, n_iter: None): 
    min(run times):    0.90s, max(run times):    0.92s, compile time:     1.1s, peak memory: 8.4e+03B, max(abs(error)): 6.8e-12, floating point ops: 1.9e+08, mem access: 3.7e+08B
(method: jax, L: 512, L_lower: 0, sampling: mw, spin: 0, L_to_nside_ratio: 2, reality: True, spmd: False, n_iter: None): 
    min(run times):     9.0s, max(run times):     9.1s, compile time:     1.6s, peak memory: 8.4e+03B, max(abs(error)): 5.4e-11, floating point ops: 8.3e+08, mem access: 1.5e+09B
```

### CODE REVIEW

File: s2fft/utils/quadrature_jax.py [hunk 1]
```diff
@@ -154,12 +154,18 @@ def cond(arg):
...
-            for j in range(1, L + 1):
+            p1 = jnp.ones_like(z)
+            p2 = jnp.zeros_like(z)
+
+            def inner_loop_step(j, p1_p2):
+                p1, p2 = p1_p2
                 p3 = p2
                 p2 = p1
                 p1 = ((2.0 * j - 1.0) * z * p2 - (j - 1.0) * p3) / j
+                return p1, p2
```
Comment (willGraham01) [inline s2fft/utils/quadrature_jax.py]:
```suggestion
            def inner_loop_step(j, p1_p2):
                p1, p2 = p1_p2
                return ((2.0 * j - 1.0) * z * p1 - (j - 1.0) * p2) / j, p1
```

Unless there's some in-place / shared memory concerns, I believe this is functionally equivalent and avoids having to trace assignments through different variables.

**Review by willGraham01 — APPROVED** [review]
One loop-function I think can be tidied up from a readability perspective, otherwise all good.

Would be interested to know if the use of a loop for the `k` and `w` related calculation is because memory concerns, or if my understanding of array operations vs loops is wrong 😅

File: s2fft/utils/quadrature_jax.py [hunk 2]
```diff
@@ -208,10 +214,12 @@ def quad_weight_dh_theta_only(theta: float, L: int) -> float:
...
     """
-    w = 0.0
-    for k in range(0, L):
+
+    def increment_weights(k, w):
         w += jnp.sin((2 * k + 1) * theta) / (2 * k + 1)
+        return w
 
+    w = jax.lax.fori_loop(0, L, increment_weights, jnp.zeros_like(theta))
     w *= 2 / L * jnp.sin(theta)
```
[the remark this answers is not in the document]
Comment (matt-graham) [reply s2fft/utils/quadrature_jax.py]: You're definitely right that this could be replaced with a vectorized expression and I suspect in terms of runtime performance this will be better in general, particularly on accelerator devices as the summation should be able to parallelized, and an $O(L)$ memory overhead would be fine. 

The type hints are a bit misleading here however as this is being called with `theta` as an $O(L)$ length 1D array rather than a float. I think to get your version to work with a vector `theta` we would need to add in some singleton dimensions to get broadcasting working (or possibly use `jax.vmap` instead)  and adjust the sum to be over a particular axis - something like

```Python
w = jnp.sin(thetas[:, None] * two_k_plus_1[None]) / two_k_plus_1[None]).sum(-1) * jnp.sin(thetas) * 2 / L
```

This would then also instantiate an intermediate array of size $O(L^2)$ which I think would start to be more of an issue memory wise (I think there are other cases where we incur this sort of memory overhead but as we are memory limited for large bandlimits we want to keep memory usage as low as possible). 

As overall this computation of the quadrature weights is not a significant component of runtime I think probably erring on the side lower memory requirements is probably the safer choice here as that is likely to be the bigger effect for large `L`.
